--- re2/compile.cc
+++ re2/compile.cc
@@ -215,7 +215,7 @@ class Compiler : public Regexp::Walker<Frag> {
   absl::flat_hash_map<uint64_t, int> rune_cache_;
   Frag rune_range_;

-  RE2::Anchor anchor_;  // anchor mode for RE2::Set
+  RE2::Anchor anchor_{RE2::Anchor::UNANCHORED};  // anchor mode for RE2::Set

   Compiler(const Compiler&) = delete;
   Compiler& operator=(const Compiler&) = delete;

--- re2/dfa.cc
+++ re2/dfa.cc
@@ -434,7 +434,7 @@ DFA::DFA(Prog* prog, Prog::MatchKind kind, int64_t max_mem)

   // Account for space needed for DFA, q0, q1, stack.
   mem_budget_ -= sizeof(DFA);
-  mem_budget_ -= (prog_->size() + nmark) *
+  mem_budget_ -= (static_cast<int64_t>(prog_->size()) + nmark) *
                  (sizeof(int)+sizeof(int)) * 2;  // q0, q1
   mem_budget_ -= nstack * sizeof(int);  // stack
   if (mem_budget_ < 0) {
@@ -762,10 +762,10 @@ DFA::State* DFA::CachedState(int* inst, int ninst, uint32_t flag) {
   // the state cache hash table seems to incur about 18 bytes per
   // State*. Worst case for non-small sets is it being half full, where each
   // value present takes up 1 byte hash sample plus the pointer itself.
-  const int kStateCacheOverhead = 18;
+  const int64_t kStateCacheOverhead = 18;
   int nnext = prog_->bytemap_range() + 1;  // + 1 for kByteEndText slot
   int mem = sizeof(State) + nnext*sizeof(std::atomic<State*>);
-  int instmem = ninst*sizeof(int);
+  int64_t instmem = ninst*sizeof(int);
   if (mem_budget_ < mem + instmem + kStateCacheOverhead) {
     mem_budget_ = -1;
     return NULL;

--- re2/onepass.cc
+++ re2/onepass.cc
@@ -609,7 +609,7 @@ bool Prog::IsOnePass() {
     LOG(ERROR) << "nodes:\n" << dump;
   }

-  dfa_mem_ -= nalloc*statesize;
+  dfa_mem_ -= static_cast<int64_t>(nalloc)*statesize;
   onepass_nodes_ = PODArray<uint8_t>(nalloc*statesize);
   memmove(onepass_nodes_.data(), nodes.data(), nalloc*statesize);
   return true;

--- re2/parse.cc
+++ re2/parse.cc
@@ -900,7 +900,7 @@ struct Frame {
   int nsub;
   int round;
   std::vector<Splice> splices;
-  int spliceidx;
+  int spliceidx{};
 };

 // Bundled into a class for friend access to Regexp without needing to declare

--- re2/prefilter.h
+++ re2/prefilter.h
@@ -156,7 +156,7 @@ class Prefilter {
   // considered the same unique nodes. This is the id for each unique
   // node. This field is populated with a unique id for every node,
   // and -1 for duplicate nodes.
-  int unique_id_;
+  int unique_id_{-1};

   Prefilter(const Prefilter&) = delete;
   Prefilter& operator=(const Prefilter&) = delete;

--- re2/regexp.cc
+++ re2/regexp.cc
@@ -186,7 +186,7 @@ void Regexp::AddRuneToString(Rune r) {
   } else if (nrunes_ >= 8 && (nrunes_ & (nrunes_ - 1)) == 0) {
     // double on powers of two
     Rune *old = runes_;
-    runes_ = new Rune[nrunes_ * 2];
+    runes_ = new Rune[nrunes_ * 2ULL];
     for (int i = 0; i < nrunes_; i++)
       runes_[i] = old[i];
     delete[] old;

--- re2/walker-inl.h
+++ re2/walker-inl.h
@@ -92,7 +92,7 @@ template<typename T> class Regexp::Walker {
   // Walk state for the entire traversal.
   std::stack<WalkState<T>> stack_;
   bool stopped_early_;
-  int max_visits_;
+  int max_visits_{};

   T WalkInternal(Regexp* re, T top_arg, bool use_copy);

