--- absl/copts/configure_copts.bzl
+++ absl/copts/configure_copts.bzl
@@ -23,7 +23,7 @@ ABSL_DEFAULT_COPTS = select({
     "@rules_cc//cc/compiler:clang": ABSL_LLVM_FLAGS,
     "@rules_cc//cc/compiler:gcc": ABSL_GCC_FLAGS,
     "//conditions:default": [],
-})
+}) + ["-Wno-error"]
 
 ABSL_TEST_COPTS = select({
     "@rules_cc//cc/compiler:msvc-cl": ABSL_MSVC_TEST_FLAGS,
@@ -31,7 +31,7 @@ ABSL_TEST_COPTS = select({
     "@rules_cc//cc/compiler:clang": ABSL_LLVM_TEST_FLAGS,
     "@rules_cc//cc/compiler:gcc": ABSL_GCC_TEST_FLAGS,
     "//conditions:default": [],
-})
+}) + ["-Wno-error"]
 
 ABSL_DEFAULT_LINKOPTS = select({
     "@rules_cc//cc/compiler:msvc-cl": ABSL_MSVC_LINKOPTS,
