--- absl/strings/internal/str_format/float_conversion.cc
+++ absl/strings/internal/str_format/float_conversion.cc
@@ -385,7 +385,7 @@ char* PrintFractionalDigitsFast(uint64_t v,
                                 int exp,
                                 size_t precision) {
   char *p = start;
-  v <<= (64 - exp);
+  v <<= (64ULL - exp);
   while (precision > 0) {
     if (!v) return p;
     *p++ = MultiplyBy10WithCarry(&v, 0) + '0';
@@ -855,7 +855,7 @@ void FormatANormalize(const HexFloatTypeParams float_traits, uint8_t *leading,
   // get interrupted by the exponent threshold.
   while (*mantissa && !(*mantissa & kHighIntBit)) {
     if (ABSL_PREDICT_FALSE(*exp - 1 < float_traits.min_exponent)) {
-      *mantissa >>= (float_traits.min_exponent - *exp);
+      *mantissa >>= (static_cast<uint64_t>(float_traits.min_exponent) - *exp);
       *exp = float_traits.min_exponent;
       return;
     }
