--- absl/synchronization/BUILD.bazel
+++ absl/synchronization/BUILD.bazel
@@ -126,7 +126,7 @@ cc_library(
     linkopts = select({
         "@rules_cc//cc/compiler:msvc-cl": [],
         "@rules_cc//cc/compiler:clang-cl": [],
-        "@rules_cc//cc/compiler:emscripten": [],
+        #"@rules_cc//cc/compiler:emscripten": [],
         "//conditions:default": ["-pthread"],
     }) + ABSL_DEFAULT_LINKOPTS,
     visibility = ["//visibility:public"],
