diff --git a/include/jemalloc/internal/jemalloc_internal_inlines_b.h b/include/jemalloc/internal/jemalloc_internal_inlines_b.h
index dad37a9..434f7af 100644
--- a/include/jemalloc/internal/jemalloc_internal_inlines_b.h
+++ b/include/jemalloc/internal/jemalloc_internal_inlines_b.h
@@ -11,6 +11,10 @@ percpu_arena_update(tsd_t *tsd, unsigned cpu) {
 	assert(have_percpu_arena);
 	arena_t *oldarena = tsd_arena_get(tsd);
 	assert(oldarena != NULL);
+	if(oldarena == NULL) {
+		return;
+	}
+
 	unsigned oldind = arena_ind_get(oldarena);
 
 	if (oldind != cpu) {
diff --git a/include/jemalloc/internal/util.h b/include/jemalloc/internal/util.h
index ecfa76b..51e8b74 100644
--- a/include/jemalloc/internal/util.h
+++ b/include/jemalloc/internal/util.h
@@ -106,11 +106,13 @@ util_prefetch_read(void *ptr) {
 #ifdef __GNUC__
 	if (config_debug) {
 		/* Enforce the "valid ptr" requirement. */
-		*(volatile char *)ptr;
+		volatile char byte = *(volatile char *)ptr;
+		(void)byte;
 	}
 	__builtin_prefetch(ptr, /* read or write */ 0, /* locality hint */ 3);
 #else
-	*(volatile char *)ptr;
+	volatile char byte = *(volatile char *)ptr;
+	(void)byte;
 #endif
 }
 
@@ -118,7 +120,8 @@ JEMALLOC_ALWAYS_INLINE void
 util_prefetch_write(void *ptr) {
 #ifdef __GNUC__
 	if (config_debug) {
-		*(volatile char *)ptr;
+		volatile char byte = *(volatile char *)ptr;
+		(void)byte;
 	}
 	/*
 	 * The only difference from the read variant is that this has a 1 as the
@@ -126,7 +129,8 @@ util_prefetch_write(void *ptr) {
 	 */
 	__builtin_prefetch(ptr, 1, 3);
 #else
-	*(volatile char *)ptr;
+	volatile char byte = *(volatile char *)ptr;
+	(void)byte;
 #endif
 }
 
diff --git a/src/bin.c b/src/bin.c
index 6bab4b2..dfe8b90 100644
--- a/src/bin.c
+++ b/src/bin.c
@@ -102,10 +102,25 @@ bin_slab_reg_alloc_batch(
 #else
 	unsigned group = 0;
 	bitmap_t g = slab_data->bitmap[group];
+
+	const size_t ngroups = bitmap_size(&bin_info->bitmap_info) >> LG_SIZEOF_BITMAP;
+
 	unsigned i = 0;
 	while (i < cnt) {
 		while (g == 0) {
-			g = slab_data->bitmap[++group];
+			group++;
+
+			assert(group < ngroups);
+			if(unlikely(group >= ngroups)) {
+				for (unsigned i = 0; i < cnt; i++) {
+					size_t regind = bitmap_sfu(slab_data->bitmap,
+								&bin_info->bitmap_info);
+					*(ptrs + i) = (void *)((uintptr_t)edata_addr_get(slab) +
+						(uintptr_t)(bin_info->reg_size * regind));
+				}
+				goto slow_fallback_done;
+			}
+			g = slab_data->bitmap[group];
 		}
 		size_t shift = group << LG_BITMAP_GROUP_NBITS;
 		size_t pop = popcount_lu(g);
@@ -129,6 +144,7 @@ bin_slab_reg_alloc_batch(
 		}
 		slab_data->bitmap[group] = g;
 	}
+slow_fallback_done:
 #endif
 	edata_nfree_sub(slab, cnt);
 }
diff --git a/src/prof_data.c b/src/prof_data.c
index 7aa047a..b66e234 100644
--- a/src/prof_data.c
+++ b/src/prof_data.c
@@ -338,6 +338,10 @@ prof_lookup(tsd_t *tsd, prof_bt_t *bt) {
 	tdata = prof_tdata_get(tsd, false);
 	assert(tdata != NULL);
 
+	if(tdata == NULL) {
+			return ret.p;
+	}
+
 	malloc_mutex_lock(tsd_tsdn(tsd), tdata->lock);
 	not_found = ckh_search(&tdata->bt2tctx, bt, NULL, &ret.v);
 	if (!not_found) { /* Note double negative! */
diff --git a/src/tcache.c b/src/tcache.c
index fe210d2..d2538e4 100644
--- a/src/tcache.c
+++ b/src/tcache.c
@@ -7,6 +7,7 @@
 #include "jemalloc/internal/safety_check.h"
 #include "jemalloc/internal/san.h"
 #include "jemalloc/internal/sc.h"
+#include "jemalloc/internal/util.h"
 
 /******************************************************************************/
 /* Data. */
@@ -1244,8 +1245,11 @@ tcache_destroy(tsd_t *tsd, tcache_t *tcache, bool tsd_tcache) {
 	 * tsd).  Manually trigger decay to avoid pathological cases.  Also
 	 * include arena 0 because the tcache array is allocated from it.
 	 */
-	arena_decay(
-	    tsd_tsdn(tsd), arena_get(tsd_tsdn(tsd), 0, false), false, false);
+	arena_t *a0 = arena_get(tsd_tsdn(tsd), 0, false);
+	if(likely(a0 != NULL)) {
+		arena_decay(tsd_tsdn(tsd), a0,
+			/* is_background_thread */ false, /* all */ false);
+	}
 
 	if (arena_nthreads_get(arena, false) == 0
 	    && !background_thread_enabled()) {
