--- a/src/arena.c
+++ b/src/arena.c
@@ -257,10 +257,25 @@ arena_slab_reg_alloc_batch(edata_t *slab, const bin_info_t *bin_info,
 #else
 	unsigned group = 0;
 	bitmap_t g = slab_data->bitmap[group];
+
+	const size_t ngroups = bitmap_size(&bin_info->bitmap_info) >> LG_SIZEOF_BITMAP;
+
 	unsigned i = 0;
 	while (i < cnt) {
 		while (g == 0) {
-			g = slab_data->bitmap[++group];
+			group++;
+
+			assert(group < ngroups);
+			if(unlikely(group >= ngroups)) {
+				for (unsigned i = 0; i < cnt; i++) {
+					size_t regind = bitmap_sfu(slab_data->bitmap,
+								&bin_info->bitmap_info);
+					*(ptrs + i) = (void *)((uintptr_t)edata_addr_get(slab) +
+						(uintptr_t)(bin_info->reg_size * regind));
+				}
+				goto slow_fallback_done;
+			}
+			g = slab_data->bitmap[group];
 		}
 		size_t shift = group << LG_BITMAP_GROUP_NBITS;
 		size_t pop = popcount_lu(g);
@@ -283,6 +298,7 @@ arena_slab_reg_alloc_batch(edata_t *slab, const bin_info_t *bin_info,
 		}
 		slab_data->bitmap[group] = g;
 	}
+slow_fallback_done:
 #endif
 	edata_nfree_sub(slab, cnt);
 }
