--- a/src/hpa.c
+++ b/src/hpa.c
@@ -707,10 +707,16 @@ hpa_alloc_batch_psset(tsdn_t *tsdn, hpa_shard_t *shard, size_t size,
 
 static hpa_shard_t *
 hpa_from_pai(pai_t *self) {
-	assert(self->alloc = &hpa_alloc);
-	assert(self->expand = &hpa_expand);
-	assert(self->shrink = &hpa_shrink);
-	assert(self->dalloc = &hpa_dalloc);
+#ifndef NDEBUG
+	self->alloc = &hpa_alloc;
+	self->expand = &hpa_expand;
+	self->shrink = &hpa_shrink;
+	self->dalloc = &hpa_dalloc;
+	assert(self->alloc);
+	assert(self->expand);
+	assert(self->shrink);
+	assert(self->dalloc);
+#endif // NDEBUG
 	return (hpa_shard_t *)self;
 }
 
