--- a/src/tcache.c
+++ b/src/tcache.c
@@ -6,6 +6,7 @@
 #include "jemalloc/internal/safety_check.h"
 #include "jemalloc/internal/san.h"
 #include "jemalloc/internal/sc.h"
+#include "jemalloc/internal/util.h"
 
 /******************************************************************************/
 /* Data. */
@@ -823,8 +824,11 @@ tcache_destroy(tsd_t *tsd, tcache_t *tcache, bool tsd_tcache) {
 	 * tsd).  Manually trigger decay to avoid pathological cases.  Also
 	 * include arena 0 because the tcache array is allocated from it.
 	 */
-	arena_decay(tsd_tsdn(tsd), arena_get(tsd_tsdn(tsd), 0, false),
-	    false, false);
+	arena_t *a0 = arena_get(tsd_tsdn(tsd), 0, false);
+	if(likely(a0 != NULL)) {
+		arena_decay(tsd_tsdn(tsd), a0, 
+			/* is_background_thread */ false, /* all */ false);
+	}
 
 	if (arena_nthreads_get(arena, false) == 0 &&
 	    !background_thread_enabled()) {
