--- snappy.cc
+++ snappy.cc
@@ -1167,9 +1167,8 @@ inline uint32_t ExtractOffset(uint32_t val, size_t tag_type) {
 #if defined(__x86_64__)
   constexpr uint64_t kExtractMasksCombined = 0x0000FFFF00FF0000ull;
   uint16_t result;
-  memcpy(&result,
-         reinterpret_cast<const char*>(&kExtractMasksCombined) + 2 * tag_type,
-         sizeof(result));
+  // NOLINTNEXTLINE
+  memcpy(&result, reinterpret_cast<const char*>(&kExtractMasksCombined) + 2 * tag_type, sizeof(result));
   return val & result;
 #elif defined(__aarch64__)
   constexpr uint64_t kExtractMasksCombined = 0x0000FFFF00FF0000ull;
@@ -1312,7 +1311,7 @@ class SnappyDecompressor {
   const char* ip_limit_;  // Points just past buffered bytes
   // If ip < ip_limit_min_maxtaglen_ it's safe to read kMaxTagLength from
   // buffer.
-  const char* ip_limit_min_maxtaglen_;
+  const char* ip_limit_min_maxtaglen_{};
   uint32_t peeked_;                  // Bytes peeked from reader (need to skip)
   bool eof_;                         // Hit end of input without an error?
   char scratch_[kMaximumTagLength];  // See RefillTag().
@@ -2171,7 +2170,7 @@ class SnappyScatteredWriter {
   // we keep track of all of the generated data as an array of blocks.
   // All of the blocks except the last have length kBlockSize.
   std::vector<char*> blocks_;
-  size_t expected_;
+  size_t expected_{};
 
   // Total size of all fully generated blocks so far
   size_t full_size_;
