--- a/src/roaring_array.c
+++ b/src/roaring_array.c
@@ -92,7 +92,7 @@ bool ra_init_with_capacity(roaring_array_t *new_ra, uint32_t cap) {
 }
 
 int ra_shrink_to_fit(roaring_array_t *ra) {
-    int savings = (ra->allocation_size - ra->size) *
+    int savings = ((int64_t)(ra->allocation_size) - ra->size) *
                   (sizeof(uint16_t) + sizeof(container_t *) + sizeof(uint8_t));
     if (!realloc_array(ra, ra->size)) {
       return 0;
