--- a/src/roaring.c
+++ b/src/roaring.c
@@ -225,13 +225,13 @@ roaring_bitmap_t *roaring_bitmap_from_range(uint64_t min, uint64_t max,
     do {
         uint32_t key = (uint32_t)min_tmp >> 16;
         uint32_t container_min = min_tmp & 0xFFFF;
-        uint32_t container_max = (uint32_t)minimum_uint64(max - (key << 16), 1 << 16);
+        uint32_t container_max = (uint32_t)minimum_uint64(max - ((uint64_t)key << 16), 1 << 16);
         uint8_t type;
         container_t *container = container_from_range(&type, container_min,
                                                container_max, (uint16_t)step);
         ra_append(&answer->high_low_container, key, container, type);
         uint32_t gap = container_max - container_min + step - 1;
-        min_tmp += gap - (gap % step);
+        min_tmp += (uint64_t)gap - ((uint64_t)gap % step);
     } while (min_tmp < max);
     // cardinality of bitmap will be ((uint64_t) max - min + step - 1 ) / step
     return answer;
@@ -1730,6 +1730,7 @@ roaring_uint32_iterator_t *roaring_copy_uint32_iterator(
     const roaring_uint32_iterator_t *it) {
     roaring_uint32_iterator_t *newit =
         (roaring_uint32_iterator_t *)roaring_malloc(sizeof(roaring_uint32_iterator_t));
+    if (newit == NULL) return NULL;
     memcpy(newit, it, sizeof(roaring_uint32_iterator_t));
     return newit;
 }
@@ -3333,7 +3334,7 @@ roaring_bitmap_t *roaring_bitmap_portable_deserialize_frozen(const char *buf) {
 
     size_t alloc_size = 0;
     alloc_size += sizeof(roaring_bitmap_t);
-    alloc_size += num_containers * sizeof(container_t*);
+    alloc_size += num_containers * (int64_t)sizeof(container_t*);
     alloc_size += num_bitset_containers * sizeof(bitset_container_t);
     alloc_size += num_run_containers * sizeof(run_container_t);
     alloc_size += num_array_containers * sizeof(array_container_t);
